WIP t6 custom map loading

LJW-Dev opened the pull request:
Notes:
- Techsets are loaded and dumped with my custom implementation, I think it should use the official techset structure but this isn't super important.
- The world asset uses a .fbx file and uses OpenFBX as it's loader
- It doesn't use any xmodels apart from the skybox, as I found them too complicated to load from the world asset. Definitely should be implemented somehow though.
- You need 6 gsc files for the link to be successful and for the map to load on bo2, but 4 of them are redundant (the _fx and _amb GSC files). This could be changed to make it simpler for users to create their custom maps, but i'm not sure if simplifying it is a good design choice.
- Maps require footstep tables, but there's no loading/dumping of them right now. I add empty footstep tables so the map will load, we could change this so users can edit the footstep tables but i don't know how this would be done.

LJW-Dev committed:
Added the ability to dump and load technique sets.
note: techniques should be dumped in seperate files from technique sets, but compiling them all into the techset works as well but is harder to read.

LJW-Dev committed:
Initial commit of BSP compilation

LJW-Dev committed:
merge extended techset functioanality

LJW-Dev committed:
WIP fix, loading is currently not working correctly

LJW-Dev committed:
Fix: technqiues weren't opened with the proper file path.

LJW-Dev committed:
Merge branch 't6-extend-techsets' into bsp-compilation

LJW-Dev committed:
fix: more incorrect paths

LJW-Dev committed:
Merge branch 't6-extend-techsets' into bsp-compilation

LJW-Dev committed:
all known bugs fixed and loaded maps work correctly now.

LJW-Dev committed:
V1 release of custom map code

LJW-Dev committed:
Failed test to make shadows work

LJW-Dev committed:
Second failed attempt at making shadows work

LJW-Dev committed:
remove GSC-Tool use

LJW-Dev committed:
Remove GSC-Tool 2

LJW-Dev committed:
Another refactor to clean up comments and code

LJW-Dev committed:
Update T6 json material loading to include JSON being passed directy instead of a stream.

LJW-Dev committed:
Another refactor to remove leacy code that isn't used anymore

LJW-Dev committed:
Merge fixup

LJW-Dev committed:
Updated custom map objloading code to match new structure.

LJW-Dev committed:
Updated Json Material loader to support passing JSON instead of a stream, and added techset loading code

LJW-Dev committed:
Merge branch 'main' into bsp-compilation

LJW-Dev committed:
FIxed techset dumping and loading to be the same, as well as added a missing image check to the BSP compiler to reduce the output prints when an image is missing.

LJW-Dev committed:
Added a default entity string if the entity file isn't given

Laupetin commented on src/ObjLoading/Game/T6/CustomMap/BinarySpacePartitionTreePreCalc.h:
```diff
@@ -0,0 +1,232 @@
+class BSPLeaf
+{
+private:
+    std::vector<Object*> objectList;
```
That should probably be a unique_ptr

Laupetin commented on src/ObjLoading/Game/T6/CustomMap/ProjectCreator.cpp:
```diff
@@ -0,0 +1,395 @@
+        indexVec.resize(indexVec.size() + num_indices);
+        for (size_t m = 0; m < num_indices; m++)
+        {
+            indexVec[currIndexVecSize + m] = (uint16_t)indices[m];
```
nit: the code mixes C style casts and C++ style casts a lot. The project generally only uses C++ style casts to be more explicit with what should happen and prevent unintended sideeffects.

Laupetin reviewed (commented):
Hey there, thank you for the PR, the feature is pretty awesome :)
I've looked through a lot of the code by now, sorry it takes so long, it's a big PR 😅 
I probably won't have that much time in the upcoming days, so i'll submit the comments I have currently.

I haven't looked at CustomMapLinker.h yet, though since I do not know that much about the BSPs yet, I probably won't be able to say about that part 😅 
So generally my comments are more related to code structure and memory management for example.

Also please forgive that I'm adding a lot of comments, it's just my style of doing code review (I'm adding a comment when I have any special thoughts about some place so it ends up being pretty verbose) and does not mean that necessarly everything needs to be addressed or must be solved in a particular way necessarly.
Also I can address a lot of the stuff as well, you don't have to do everything 😛

Again, thanks for putting the work in :D Looking forward to this feature and how it evolves

LJW-Dev commented on src/Common/Game/T6/T6.h:
```diff
@@ -158,6 +158,100 @@ namespace T6
+        
+        size_t modelCount;
+        customMapModel* models;
+    };
```
Would the separate header file be in the objcommon project or the objloading project?

Laupetin replied:
Generally at the same location as the cpp file.
And for those -> if it is used in the loading code only -> ObjLoading
If the dumping code uses it as well -> ObjCommon

LJW-Dev commented on src/Common/Game/T6/T6_Assets.h:
```diff
@@ -1278,6 +1278,30 @@ namespace T6
+        unsigned __int16 zStart;
+        unsigned __int16 zCount;
+        unsigned int firstEntry;
+        GfxLightGridUnk unk;
```
True, I can see it being a lookup table. I'll update the code to try match this better.

LJW-Dev committed:
WIP: Updating code to follow laupentin's code review.
Done:
- Moved custom map structures to their own objcommon header file
- Updated GfxLightGridRow struct
- Reverted shader_bin file path
- Renamed Project Creator to BSP Creator
- Removed model loading from BSP creator
- Cleaned up BSP Creator and updated the names of structs
WIP:
- Update BSP calculation code to be more readable and use unique/shared ptrs

LJW-Dev committed:
WIP:
- Converted BSP tree generation to use unique ptrs
- Updated the linker to use BSP tree unique ptrs and use the new structure of custom map structs

LJW-Dev committed:
WIP:
- Updated names to allign with the BSP naming structure
- Updated constants file and assoicated uses

LJW-Dev committed:
WIP:
- Updated file structure to use BSP naming scheme
- Re-wrote BSP creator to use c++ more efficiently

LJW-Dev committed:
WIP: Prepared project for updating CustomMapLinker.h

LJW-Dev committed:
WIP: Converted custom map linker into different asset files

LJW-Dev committed:
WIP: Finished updating BSP linker to use seperate files for each asset.

LJW-Dev committed:
Refactor to improve C++ and safe code use

LJW-Dev committed:
Refactored clipmap generation, further refactored other files to better use c++

LJW-Dev committed:
Refactored GFX world loading

LJW-Dev committed:
Refactored to make x64 builds succeed

LJW-Dev committed:
Formatted all BSP loading files using clang.

LJW-Dev committed:
Updated missing image and colour only image asset names to use materials loaded by the game.

LJW-Dev committed:
CLANG formatted ObjLoaderT6

LJW-Dev committed:
Added map type metadata parsing to zone file

LJW-Dev committed:
Added the zone definition to the parameters of ConfigureCreatorCollection. ObjLoaderT6 has also been changed to conditionally add the BSP creator if the map type isn't NONE.

LJW-Dev committed:
Updated BSP linker to load assets on the FinalizeZone step.

LJW-Dev committed:
The FinalizeZone step can now fail and the zone link will not ouput a fastfile if the FinalizeZone step fails.

LJW-Dev committed:
Assets with a name starting with ',' will now be always loaded as a default dependency.

LJW-Dev committed:
The linker should continue with building the fastfile if the asset list is empty and the map type isn't NONE.

LJW-Dev committed:
Refactored files to use CLANG formatting and linux compatabillty.

LJW-Dev committed:
Added ufbx to submodules.

LJW-Dev committed:
Added premake files for ufbx

LJW-Dev committed:
Updated ufbx premake lua to properly add it to the project

LJW-Dev committed:
Added ufbx as an ObjLoading include

LJW-Dev committed:
WIP: Updated premake scripts, removed non-submodule ufbx code, ported BSPCreator to use the submodule. OAT link still fails due to unknown premake issue.

LJW-Dev committed:
Fixed missing parameter in the ufbx premake script causing the build to fail.

LJW-Dev committed:
Undo changes to JsonMaterialLoader as they are no longer needed.

LJW-Dev committed:
Merge branch 'main' into bsp-compilation

LJW-Dev committed:
Moved coordinate conversion to BSP creator instead of the BSP linker, and added a console print of the maximum number of objects in a collision leaf.

LJW-Dev committed:
Extended BSP creation to support X, Y and Z splitting and simplified clipmap BSP generation.

LJW-Dev committed:
Fixed a crash related to AABB Tree children counts exceeding the maximum limit

LJW-Dev committed:
Moved the re-ordering of indices to the BSP creator instead of the BSP asset linker.

Laupetin commented on src/ObjLoading/Game/T6/BSP/Linker/BSPLinker.h:
```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+#include "../BSP.h"
```
The project does not use backwards directory traversal for paths 😅 

`Game/T6/BSP/BSP.h` is prefered

Laupetin commented on src/ObjLoading/Game/T6/BSP/LoaderBSP_T6.cpp:
The rest of the files that are game specific do not have an underscore before the game suffix.

I guess you made it because otherwise you would not have a separator between the T of T6 and the P of BSP?
In that case maybe `LoaderBspT6` would be an option?

Laupetin commented on src/ObjLoading/Game/T6/BSP/BSPUtil.cpp:
```diff
@@ -0,0 +1,173 @@
+
+    size_t BSPUtil::allignBy128(size_t size)
+    {
+        return ((size + 127) & 0xFFFFFF80);
```
There's a utility method for it already in `Utils/Alignment.h`:

```cpp
utils::Align(size, 128uz);
```

Laupetin commented on src/ObjLoading/Game/T6/BSP/BSPUtil.cpp:
```diff
@@ -0,0 +1,173 @@
+{
+    std::string BSPUtil::getFileNameForBSPAsset(std::string& assetName)
+    {
+        return std::format("BSP/{}", assetName);
```
Folders are generally lowercase. On Windows this does not make a difference, on Linux it does.

Laupetin commented on src/ObjLoading/Game/T6/BSP/BSPUtil.cpp:
```diff
@@ -0,0 +1,173 @@
+            AABBMins.z = newAABBMins.z;
+
+        if (newAABBMaxs.z > AABBMaxs.z)
+            AABBMaxs.z = newAABBMaxs.z;
```
These can be changed to
```cpp
        AABBMins.x = std::min(AABBMins.x, newAABBMins.x);
        AABBMaxs.x = std::max(AABBMaxs.x, newAABBMaxs.x);

        AABBMins.y = std::min(AABBMins.y, newAABBMins.y);
        AABBMaxs.y = std::max(AABBMaxs.y, newAABBMaxs.y);

        AABBMins.z = std::min(AABBMins.z, newAABBMins.z);
        AABBMaxs.z = std::max(AABBMaxs.z, newAABBMaxs.z);
```

which is a bit more readable imo 😅 

Adjusted in a8456f88a49fae340446833234f33df1c2274e5b

Laupetin commented on src/ObjLoading/Game/T6/BSP/LoaderBSP_T6.h:
```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+#include "Asset/IAssetCreator.h"
+#include "Game/T6/T6.h"
```
nit: This header seems unused

adjusted in a8fc10df74a0acb264894edbf7988fde88ea0d85

Laupetin commented on src/ObjLoading/Game/T6/BSP/BSPCreator.cpp:
```diff
@@ -0,0 +1,276 @@
+
+namespace
+{
+    using namespace BSP;
```
Usually you only do `using namespace` on global level since this technically embed one namespace into another which can get a little weird, especially if that namespace it is embedded into is named 😅 

adjusted in 08a8a5cd077ae217b1abeac4b0c8889f2b46d8b2

Laupetin commented on src/ObjLoading/Game/T6/BSP/LoaderBSP_T6.cpp:
```diff
@@ -0,0 +1,60 @@
+        {
+        }
+
+        std::optional<asset_type_t> GetHandlingAssetType() const override
```
nit: Not _that_ important but can be helpful sometimes: There is a c++ feature called [`nodiscard`](https://en.cppreference.com/w/cpp/language/attributes/nodiscard) that raises a compiler warning if the return value of a function is not used.

Helpful at functions that do nothing other than return a value.
Generally mostly applicable when a method is static or const.

adjusted in: 8f08aca15a2c13438e26cfe25e8aa17087ee6e3a

Laupetin commented on src/ObjLoading/Game/T6/BSP/Linker/BSPLinker.cpp:
```diff
@@ -0,0 +1,106 @@
+        if (m_context.LoadDependency<AssetScript>("clientscripts/mp/" + bsp->name + "_amb.csc") == nullptr)
+            return false;
+        if (m_context.LoadDependency<AssetScript>("clientscripts/mp/" + bsp->name + "_fx.csc") == nullptr)
+            return false;
```
nit: the codebase generally tries to use std::format instead of string concatination

adjusted in: 5e06df6286f30de54027f115c08c94e986bfe73c

Laupetin commented on src/ObjLoading/Game/T6/BSP/Linker/GfxWorldLinker.cpp:
```diff
@@ -0,0 +1,772 @@
+        gfxWorld->outdoorLookupMatrix[3].w = 1.0f;
+
+        std::string outdoorImageName = std::string("$outdoor");
+        auto outdoorImageAsset = m_context.LoadDependency<AssetImage>(outdoorImageName);
```
nit: since `LoadDependency` takes in a `const std::string&`, you can pass in a literal value directly instead of having to make a temporary variable that you only use once.

adjusted in: 049fa2b8ae999aabfe5f91355ce4b8a8eb692f6c

Laupetin commented on src/ObjLoading/Game/T6/BSP/BSP.h:
```diff
@@ -0,0 +1,186 @@
+    // These values are hardcoded ingame and will break the map if they are changed
+    namespace BSPGameConstants
+    {
+        constexpr unsigned int MAX_COLLISION_VERTS = UINT16_MAX;
```
nit: when there is a c++ way of doing things, i often prefer it over the C way 😅 

Numeric limits in c++ are defined via [`std::numeric_limits`](https://en.cppreference.com/w/cpp/types/numeric_limits.html).
While it may have the small benefit of always being typed correctly by default, it is mostly a personal taste thing.

adjusted in: 27c3c7dccb1069d7a7e9d26645261cb5d9c0482d

Laupetin commented on src/ObjLoading/Game/T6/BSP/BSPUtil.h:
```diff
@@ -0,0 +1,24 @@
+        static void matrixTranspose3x3(const vec3_t* in, vec3_t* out);
+        static vec3_t convertStringToVec3(std::string& str);
+        static std::string convertVec3ToString(vec3_t& vec);
+    };
```
As I mentioned in a previous comment as well, in C++ it is not necessary to put every method in a class unlike object oriented languages like C#, so it is not necessary to bundle methods.

The instrument of choice is namespaces here 😄 Since all of these methods are static.

adjusted in: 4b1e81056c491c23bfe21335b43c7f07c3ffbb40

Laupetin commented on src/ObjLoading/Game/T6/BSP/BSPUtil.h:
```diff
@@ -0,0 +1,24 @@
+    class BSPUtil
+    {
+    public:
+        static std::string getFileNameForBSPAsset(std::string& assetName);
```
nit: Methods are generally named PascalCase in the codebase

adjusted in: 3d1fa7c6d9d80366143938b0f1910f71033066c2

Laupetin reviewed (commented):
I took a look at the entire code again this time and added a few comments.

As I mentioned before, **a lot** is just nitpicks and probably personal taste and since I was already on the branch, I just adjusted them quickly.
I added commits, split for each comment so any of them can be dropped again upon request.

If you don't like some of the changes I made, please tell me in the discussion 😅

LJW-Dev committed:
wip: map file format now uses gltf instead of fbx

LJW-Dev committed:
Vertex position is now transformed by node transform.

LJW-Dev committed:
Added tangent and binormal calculation, corrected scale, rotate and translate order of operations.

LJW-Dev committed:
Extended and fixed various GFXWorld linking code

LJW-Dev committed:
Minor clipmap improvements

LJW-Dev committed:
Minor improvements to GFX linker

LJW-Dev committed:
Bux fixes for the clipmap linker

LJW-Dev committed:
Clipmap linking improvements and bug fix

LJW-Dev committed:
Materials now have their own vector and are acessed via indexes

LJW-Dev committed:
WIP unique tree for each material type

LJW-Dev committed:
Fixed various warning messages

LJW-Dev committed:
Minor fixes

LJW-Dev committed:
Fixed incorrect order of operations during vertex matrix creation

LJW-Dev committed:
Clipmap linker refactor, and AABB trees are seperated by material now.

LJW-Dev committed:
feat: WIP add lights to the scene

LJW-Dev committed:
GLTF loader now has support for the punctual lights extension

LJW-Dev committed:
Spot lights now are fully implemented, with omni and directional not working.

LJW-Dev committed:
Extended GLTF classes to support: material colours, and images buffer view index / image type.

LJW-Dev committed:
BSP creator now reads colour data from materials and applies them to the vertices. Minor GFXWorld fixups and extensions to accomodate the changes.

LJW-Dev committed:
Refactored light direction calculation.

LJW-Dev committed:
BSP creator now loads seperate graphics and collision data.

LJW-Dev committed:
feat: add gltf extras to node schema

LJW-Dev committed:
feat: spawn points are now defined in the GLTF file using custom properties. Initial xmodel implementation also included.

LJW-Dev committed:
fix to bsp materials so they require a name

LJW-Dev committed:
feat: load xmodels from GLTF and link them, xmodel collision not working yet.

LJW-Dev committed:
fixed error handling for xmodels and materials, and updated xmodel linker to check if an xmodel has valid collision.

LJW-Dev committed:
feat: surface type and flags can now be set through the GLTF file

LJW-Dev committed:
feat: moved surface type to material extras, and surface flags to the node extras

LJW-Dev committed:
fix: gfx linker was not drawing the last surface in the surface array

LJW-Dev committed:
feat: clipmap linker now has support for maps smaller than the minimum BSP node size

LJW-Dev committed:
fix: clipmap AABB generation skipped first object

LJW-Dev committed:
feat: Added console print when no spawns were found

LJW-Dev committed:
feat: gfxworld now uses surface flags for surfaces

LJW-Dev committed:
feat: xmodles have toggleable shadows

LJW-Dev committed:
feat: pathnodes are added through the GLTF file

LJW-Dev committed:
chore: replace "new" keyword with correct memory allocation

LJW-Dev committed:
feat: set basic lightmap for xmodels

LJW-Dev committed:
chorse: groundwork for implementing zombies maps

LJW-Dev committed:
Addition of zombie spawns, player spawns and zones for zombies. Clipmap AABB generation largely modified.

LJW-Dev committed:
Zombies GLTF

LJW-Dev committed:
chore: various minor additions

LJW-Dev committed:
Minor additions to BSP Creator

LJW-Dev committed:
rename castSunShadow flag back to onlyCastShadow

LJW-Dev committed:
chore: Remove unused dependency ufbx

LJW-Dev committed:
fix: incorrect flag name

LJW-Dev committed:
refactor: better implementation of static and script surfaces

LJW-Dev committed:
chore: removed unused code comments

LJW-Dev committed:
feat: zone volumes now use brushes instead of terrain to define their boundaries

LJW-Dev committed:
feat: trigger_use and trigger_multiple can now be loaded through GLTF

LJW-Dev committed:
feat: fully customisable entities through gltf

LJW-Dev committed:
fix: uppercase flag name

LJW-Dev committed:
refactor: minor improvements to bsp creation

LJW-Dev committed:
feat: heirarchical transforms are now implemented

LJW-Dev committed:
fix: slight update to lightgrid after new info found

LJW-Dev committed:
chore: cleanup and minor addition

LJW-Dev committed:
feat: script_brushmodel now works correctly

LJW-Dev committed:
feat: minor bug fixes

LJW-Dev committed:
fix: always make an a material a textured one

LJW-Dev committed:
feat: Loading sun data through GLTF

LJW-Dev committed:
feat: proper imlementation of spot lights

LJW-Dev committed:
feat: Minor fixups and improvements

LJW-Dev committed:
feat: removed ents file, worldspawn and intermission are added through gltf nodes now

LJW-Dev committed:
feat: removed outdoor imagem dependency

LJW-Dev committed:
feat: proper implementation of GFX models and refactor of node ignore/allow

LJW-Dev committed:
feat: use xmodels to calculate skinnedverts

LJW-Dev committed:
refactor: remove extra classes from BSP.h and move vars to their respective files

LJW-Dev committed:
refactor: simplified brushmodels - gfx uses tris, col uses brushes only

LJW-Dev committed:
feat: fixed models flickering and not being drawn due to skinned verts

LJW-Dev committed:
chore: updated default images to look nicer ingame

LJW-Dev committed:
fix: better ignore/allow for meshes and classes when loading gltf

LJW-Dev committed:
feat: better filtering for xmodels

LJW-Dev committed:
chore: refactored node filtering

LJW-Dev committed:
chore: added many bound checks to prevent integer overflows and game imposed limits, beautified code, added comments where logic wasn't very obvious.

asajid2-cell commented:
Hey! I’ve been testing the T6 custom-map branch against Plutonium and got a generated zm_cosmodrome map to reach live Zombies gameplay/HUD. I'm still working on understanding the engine; however, currently, I've found two small upstreamable issues:

1. FBX CullingOff was not honored for generated GfxWorld surfaces.
2. T6 generated world surfaces can silently accept model-material families like mc/*, but runtime rendering only became correct when using a wpc/* world material.

I have focused tests and runtime proof screenshots. Is there any way to contribute?

LJW-Dev commented:
Right now i'm moving to using GLTF as i found it to not have these issues that FBX has had, hopefully switching over would fix the 1st issue (the branch is updated to use gltf   ##now).

I agree a warning would be good when it sees not a world material, but when custom materials become possible i doubt many map makers would use the same wpc/wc/etc naming scheme so a full error would be too much. Maybe there's a way to tell if the material will/won't render apart from their naming scheme? 

Feel free to add a pull request to the branch if you want to add changes though.

asajid2-cell commented:
Sounds good, I’ll keep any future PR small and re-check against the current GLTF path first.

For context, my local fork/runtime lane can get a generated zm_cosmodrome into Plutonium ZM gameplay, including player spawn/movement and basic zombie melee damage, but it is still very early and has local probe/staging scaffolding. It was also based on your earlier custom-map work, so I’m now trying to line up with the current branch instead of guessing.

I tried a quick sanity check on the current bsp-compilation head with a minimal GLTF test map, but hit an early runtime crash before InitGame, so I may be misunderstanding the expected source setup.

Is there a specific current commit/example GLTF you consider the known-good in-game smoke test? I can base any testing/PR work on that.
